qiskit/algorithms/gradients/finite_diff_estimator_gradient.py

```diff
@@ -0,0 +1,127 @@
+        super().__init__(estimator, **run_options)
+
+    def _run(
+        self,
+        circuits: Sequence[QuantumCircuit],
+        observables: Sequence[BaseOperator | PauliSumOp],
+        parameter_values: Sequence[Sequence[float]],
+        partial: Sequence[Sequence[Parameter]] | None = None,
+        **run_options,
+    ) -> EstimatorGradientJob:
+        partial = partial or [[] for _ in range(len(circuits))]
+        gradients = []
+        status = []
+        for circuit, observable, parameter_values_, partial_ in zip(
+            circuits, observables, parameter_values, partial
+        ):
+            index = self._circuit_ids.get(id(circuit))
+            if index is not None:
+                circuit_index = index
+            else:
+                # if the given circuit is  a new one, make base parameter values for + and - epsilon
+                circuit_index = len(self._circuits)
+                self._circuit_ids[id(circuit)] = circuit_index
+                self._base_parameter_values_dict[
+                    circuit_index
+                ] = make_fin_diff_base_parameter_values(circuit, self._epsilon)
+                self._circuits.append(circuit)
+
+            circuit_parameters = self._circuits[circuit_index].parameters
+            base_parameter_values_list = []
+            gradient_parameter_values = np.zeros(len(circuit_parameters))
+
+            # a parameter set for the partial option
+            parameters = partial_ or circuit_parameters
+            param_set = set(parameters)
+
+            result_index = 0
+            result_index_map = {}
+            # bring the base parameter values for parameters only in the partial parameter set.
+            for i, param in enumerate(circuit_parameters):
+                gradient_parameter_values[i] = parameter_values_[i]
+                if param in param_set:
+                    base_parameter_values_list.append(
+                        self._base_parameter_values_dict[circuit_index][i * 2]
+                    )
+                    base_parameter_values_list.append(
+                        self._base_parameter_values_dict[circuit_index][i * 2 + 1]
+                    )
+                    result_index_map[param] = result_index
+                    result_index += 1
+            # add the given parameter values and the base parameter values
+            gradient_parameter_values_list = [
+                gradient_parameter_values + base_parameter_values
+                for base_parameter_values in base_parameter_values_list
+            ]
+            gradient_circuits = [self._circuits[circuit_index]] * len(
+                gradient_parameter_values_list
+            )
+            observable_list = [observable] * len(gradient_parameter_values_list)
+
+            job = self._estimator.run(
+                gradient_circuits, observable_list, gradient_parameter_values_list, **run_options
+            )
+            results = job.result()
```

Comment:
Hi! We have already discussed this offline, but I thought it could be useful to leave a comment here too. If the goal of this `run` method is to return an async job, I think that we cannot call `job.result()`, but instead build a job class similar to [primitive_job](https://github.com/Qiskit/qiskit-terra/blob/14a0aaf8364e47b5c24761bcd48a7d5252ed2a8b/qiskit/primitives/primitive_job.py) that allows to submit and return results asynchronously. I wonder where does it make sense to "draw the line" to support async. Should algorithms return jobs too?  If we wanted to build an algorithm that uses gradients and can also be used in async mode, then we would have to again have these submit/return result methods, and so on so forth.

Reply:
Based on the discussion, I will use `evaluate` instead of `run`. Also, they return a result class instead of a job class.